File: legacy-dotnet/Services/DataUsage.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

namespace PrayerTray.Services;

/// <summary>Daily traffic accumulator persisted to usage.json (kept ~90 days). Tick ~1/s.</summary>
static class DataUsage
{
    public record Day(long Rx, long Tx);

    static Dictionary<string, Day> _days = new();
    static readonly Dictionary<string, (long rx, long tx)> _base = new();
    static long _lastTick, _lastSave;
    static bool _dirty;
    static string? _ifaceId;

    const int KeepDays = 90;

    static string FilePath => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
        "PrayerTray", "usage.json");

    public static void SetInterface(string? id)
    {
        if (id != _ifaceId) { _ifaceId = id; _base.Clear(); }
    }

    public static void Load()
    {
        try
        {
            if (File.Exists(FilePath))
                _days = JsonSerializer.Deserialize<Dictionary<string, Day>>(File.ReadAllText(FilePath)) ?? new();
        }
        catch { _days = new(); /* corrupt store -> start fresh */ }
    }

    public static void Tick()
    {
        long now = Environment.TickCount64;
        // A gap (sleep, tracking toggled off) means the baselines are stale; re-prime
        // rather than attribute the whole untracked window to this one tick.
        if (_lastTick > 0 && now - _lastTick > 10_000) _base.Clear();
        _lastTick = now;

        var (dr, dt) = NetSpeed.Delta(_base, _ifaceId);
        if (dr > 0 || dt > 0)
        {
            string day = DateTime.Now.ToString("yyyy-MM-dd");
            var cur = _days.TryGetValue(day, out var d) ? d : new Day(0, 0);
            _days[day] = new Day(cur.Rx + dr, cur.Tx + dt);
            _dirty = true;
        }

        if (_dirty && now - _lastSave >= 300_000) Save();
    }

    public static (long rx, long tx) Today() =>
        _days.TryGetValue(DateTime.Now.ToString("yyyy-MM-dd"), out var d) ? (d.Rx, d.Tx) : (0, 0);

    /// <summary>History newest-first, including today.</summary>
    public static List<(string date, long rx, long tx)> History()
    {
        var list = new List<(string, long, long)>(_days.Count);
        foreach (var (date, d) in _days) list.Add((date, d.Rx, d.Tx));
        list.Sort((a, b) => string.CompareOrdinal(b.Item1, a.Item1));
        return list;
    }

    public static void Reset()
    {
        _days.Clear();
        _dirty = true;
        Save();
    }

    public static void Flush() { if (_dirty) Save(); }

    static void Save()
    {
        _lastSave = Environment.TickCount64; // even on failure, retry at the normal cadence
        try
        {
            string cutoff = DateTime.Now.AddDays(-KeepDays).ToString("yyyy-MM-dd");
            var stale = new List<string>();
            foreach (var k in _days.Keys) if (string.CompareOrdinal(k, cutoff) < 0) stale.Add(k);
            foreach (var k in stale) _days.Remove(k);

            Directory.CreateDirectory(Path.GetDirectoryName(FilePath)!);
            File.WriteAllText(FilePath, JsonSerializer.Serialize(_days));
            _dirty = false;
        }
        catch { /* best effort; retried on the next interval */ }
    }

    /// <summary>"512 B" / "3.25 KB" / "1.2 GB" — invariant digits, 3 significant figures.</summary>
    public static string Size(long bytes)
    {
        if (bytes < 1024) return $"{bytes} B";
        double v = bytes / 1024.0;
        if (v < 1024) return $"{Num(v)} KB";
        v /= 1024.0;
        if (v < 1024) return $"{Num(v)} MB";
        v /= 1024.0;
        if (v < 1024) return $"{Num(v)} GB";
        return $"{Num(v / 1024.0)} TB";
    }

    static string Num(double v) => v < 10 ? v.ToString("0.00") : v < 100 ? v.ToString("0.0") : v.ToString("0");
}

File: rust/src/services/data_usage.rs
//! Daily traffic accumulator persisted to usage.json (kept ~90 days). Tick ~1/s.
//! Port of C# Services/DataUsage.cs; JSON-compatible with the v1.x file.

use crate::config::AppConfig;
use crate::native::net::IfRow;
use crate::native::time::{day_key_offset, tick_count64, today_key};
use crate::services::net_speed;
use serde::{Deserialize, Serialize};
use std::collections::{BTreeMap, HashMap};
use std::path::PathBuf;

const KEEP_DAYS: i64 = 90;
const AUTOSAVE_MS: u64 = 300_000;

#[derive(Clone, Copy, Default, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct Day {
    pub rx: u64,
    pub tx: u64,
}

pub struct DataUsage {
    days: BTreeMap<String, Day>,
    baseline: HashMap<String, (u64, u64)>,
    last_tick: u64,
    last_save: u64,
    dirty: bool,
}

fn file_path() -> PathBuf {
    AppConfig::dir().join("usage.json")
}

impl Default for DataUsage {
    fn default() -> Self {
        Self::new()
    }
}

impl DataUsage {
    pub fn new() -> Self {
        Self {
            days: BTreeMap::new(),
            baseline: HashMap::new(),
            last_tick: 0,
            last_save: 0,
            dirty: false,
        }
    }

    pub fn load(&mut self) {
        self.days = std::fs::read_to_string(file_path())
            .ok()
            .and_then(|t| serde_json::from_str(&t).ok())
            .unwrap_or_default();
    }

    pub fn tick(&mut self, rows: &[IfRow]) {
        let now = tick_count64();
        // A gap (sleep, tracking toggled off) means the baselines are stale; re-prime
        // rather than attribute the whole untracked window to this one tick.
        if self.last_tick > 0 && now.saturating_sub(self.last_tick) > 10_000 {
            self.baseline.clear();
        }
        self.last_tick = now;

        let (dr, dt) = net_speed::delta(&mut self.baseline, rows);
        if dr > 0 || dt > 0 {
            let day = today_key();
            let e = self.days.entry(day).or_default();
            e.rx += dr;
            e.tx += dt;
            self.dirty = true;
        }

        if self.dirty && now.saturating_sub(self.last_save) >= AUTOSAVE_MS {
            self.save();
        }
    }

    pub fn today(&self) -> (u64, u64) {
        self.days.get(&today_key()).map(|d| (d.rx, d.tx)).unwrap_or((0, 0))
    }

    /// History newest-first, including today (BTreeMap iterates ascending, so reverse).
    pub fn history(&self) -> Vec<(String, u64, u64)> {
        self.days.iter().rev().map(|(k, d)| (k.clone(), d.rx, d.tx)).collect()
    }

    pub fn reset(&mut self) {
        self.days.clear();
        self.dirty = true;
        self.save();
    }

    pub fn flush(&mut self) {
        if self.dirty {
            self.save();
        }
    }

    fn save(&mut self) {
        self.last_save = tick_count64(); // even on failure, retry at the normal cadence
        let cutoff = day_key_offset(KEEP_DAYS);
        self.days.retain(|k, _| k.as_str() >= cutoff.as_str());
        let _ = std::fs::create_dir_all(AppConfig::dir());
        if let Ok(json) = serde_json::to_string(&self.days) {
            if std::fs::write(file_path(), json).is_ok() {
                self.dirty = false;
            }
        }
    }
}

/// "512 B" / "3.25 KB" / "1.2 GB" — 3 significant figures.
pub fn size(bytes: u64) -> String {
    if bytes < 1024 {
        return format!("{bytes} B");
    }
    let v = bytes as f64 / 1024.0;
    if v < 1024.0 {
        return format!("{} KB", num(v));
    }
    let v = v / 1024.0;
    if v < 1024.0 {
        return format!("{} MB", num(v));
    }
    let v = v / 1024.0;
    if v < 1024.0 {
        return format!("{} GB", num(v));
    }
    format!("{} TB", num(v / 1024.0))
}

fn num(v: f64) -> String {
    if v < 10.0 {
        format!("{v:.2}")
    } else if v < 100.0 {
        format!("{v:.1}")
    } else {
        format!("{v:.0}")
    }
}
